File: java/year2021/month3/no283/MoveZeroes.java
package year2021.month3.no283;

import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class MoveZeroes {
    public static void main(String[] args) {
        int[] nums1 = {0, 1, 0, 3, 12};
        int[] nums2 = {0, 1, 0, 3, 0};
        int[] nums3 = {1};
        moveZeroes1(nums1);
        moveZeroes1(nums2);
        moveZeroes1(nums3);
        disp(nums1);
        disp(nums2);
        disp(nums3);
    }

    private static void moveZeroes1(int[] nums) {
//        Partition in Quick Sort
//        [0, slow)非零，[slow, right)必为零，[right, n)未知
        int slow = 0;
        int fast = 0;
        while (fast < nums.length) {
            if (nums[fast] != 0) {
                swap(nums, slow, fast);
                slow++;
            }
            fast++;
        }
    }

    private static void swap(int[] nums, int x, int y) {
        int temp = nums[x];
        nums[x] = nums[y];
        nums[y] = temp;
    }

    public static void moveZeroes(int[] nums) {
        int slow = -1;
        int fast = 0;
        while (fast < nums.length) {
            if (nums[fast] != 0) {
                slow++;
                nums[slow] = nums[fast];
            }
            fast++;
        }
        while (slow < nums.length - 1) {
            nums[++slow] = 0;
        }
    }

    private static void disp(int[] nums) {
        System.out.println(IntStream.of(nums).boxed().collect(Collectors.toList()));
    }
}

/*
* 给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。

示例:

输入: [0,1,0,3,12]
输出: [1,3,12,0,0]
说明:

必须在原数组上操作，不能拷贝额外的数组。
尽量减少操作次数。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/move-zeroes
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month5/no283/MoveZeroes.go
package main

import "fmt"

func main() {
	nums1 := []int{0, 1, 0, 3, 12}
	nums2 := []int{0}
	moveZeroes(nums1) // [1, 3, 12, 0, 0]
	moveZeroes(nums2) // [0]
	println(fmt.Sprintf("%v", nums1))
	println(fmt.Sprintf("%v", nums2))
}

/*
Two Pointers, times: O(n), space: O(1)
[0, slow) => 非0
[fast, len) => 0
*/
func moveZeroes(nums []int) {
	slow, fast := 0, 0
	for fast < len(nums) {
		if nums[fast] != 0 {
			nums[slow], nums[fast] = nums[fast], nums[slow]
			slow++
		}
		fast++
	}
}

/*
给定一个数组 nums，编写一个函数将所有 0 移动到数组的末尾，同时保持非零元素的相对顺序。
请注意 ，必须在不复制数组的情况下原地对数组进行操作。

示例 1:
输入: nums = [0,1,0,3,12]
输出: [1,3,12,0,0]

示例 2:
输入: nums = [0]
输出: [0]

提示:
1 <= nums.length <= 104
-231 <= nums[i] <= 231 - 1
进阶：你能尽量减少完成的操作次数吗？
*/
